diesel/src/mysql/connection/bind.rs

```diff
@@ -40,6 +71,56 @@ impl Binds {
         &mut self.mysql_binds
     }
 
+    /// The caller of this function must immediately check for errors
+    /// after return
+    pub fn populate_dynamic_buffers(&mut self, stmt: *mut ffi::MYSQL_STMT) {
+        use std::cmp::min;
+
+        for (i, buffer) in self.data.iter_mut().enumerate() {
+            if self.errors.as_ref().unwrap()[i] != 0 {
+                let written_bytes = buffer.len();
+                let needed_capacity = self.lengths[i] as usize;
+                buffer.reserve(needed_capacity - written_bytes);
+                self.mysql_binds[i].buffer = buffer.as_mut_ptr() as *mut libc::c_void;
+                mem::swap(&mut self.lengths[i], &mut self.mysql_binds[i].buffer_length);
+                let load_result = unsafe {
+                    ffi::mysql_stmt_fetch_column(
+                        stmt,
+                        &mut self.mysql_binds[i],
+                        i as libc::c_uint,
+                        written_bytes as libc::c_ulong,
+                    )
+                };
+                if load_result != 0 {
+                    return;
+                }
+            }
+
+            let actual_bytes_in_buffer = min(
+                self.mysql_binds[i].buffer_length,
+                self.lengths[i],
+            );
+            unsafe { buffer.set_len(actual_bytes_in_buffer as usize) }
```

Comment:
`set_len` can never leak memory. It only affects whether destructors are run for the values in the vec. In this case, the values are `u8` so it's meaningless. The amount of memory used is defined by capacity, which we don't touch. I don't want to use `truncate` or `shrink_to_fit`, because we'll likely just have to re-allocate on the next row. This code re-uses the buffer for each query run (deallocating after the last row is processed)